File: 152.maximum-product-subarray.js
// @leet start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
	var min = nums[0];
	var max = nums[0];
	var answer = nums[0];
	for (let i = 1; i < nums.length; i++) {
		var temp = Math.min(nums[i], Math.min(nums[i] * max, nums[i] * min));
		max = Math.max(nums[i], Math.max(nums[i] * max, nums[i] * min));
		min = temp;
		answer = Math.max(answer, max);
	}
	return answer;
};
// @leet end

File: 152.maximum-product-subarray.cpp
// @leet start
#include <algorithm>
#include <vector>
class Solution {
public:
  int maxProduct(vector<int> &nums) {
    int maximum = nums[0];
    int minimum = nums[0];
    int answer = nums[0];
    for (int i = 1; i < nums.size(); i++) {
      int temp =
          std::min(nums[i], std::min(nums[i] * maximum, nums[i] * minimum));

      maximum =
          std::max(nums[i], std::max(nums[i] * maximum, nums[i] * minimum));
			minimum = temp;
			answer = std::max(answer, maximum);
    }
		return answer;
  }
};
// @leet end
